lib/metrics/context.js

```diff
@@ -43,6 +52,60 @@ module.exports = {
 
     return data
   }
+
+  function validate(request) {
+
+    var metadata = request.payload.metricsContext
+    if (!metadata) {
+      return logInvalidContext(request, 'missing context')
+    }
+    if (!metadata.flowId) {
+      return logInvalidContext(request, 'missing flowId')
+    }
+    if (!metadata.flowBeginTime) {
+      return logInvalidContext(request, 'missing flowBeginTime')
+    }
+
+    if (Date.now() - metadata.flowBeginTime > config.metrics.flow_id_expiry) {
+      return logInvalidContext(request, 'expired flowBeginTime')
+    }
+
+    // The flowId includes a HMAC of some additional information about the
+    // user-agent, as a simple way to check it's from the right place without
+    // having to share state between the content-server and auth-server.
+    var flowSignature = new Buffer(metadata.flowId.substr(32, 64), 'hex')
+    var key = config.metrics.flow_id_key
+    var expectedSignature = crypto.createHmac('sha256', key)
+      .update([
+        metadata.flowId.substr(0, 32),
+        metadata.flowBeginTime.toString(16),
+        request.headers['user-agent']
+      ].join('\n'))
+      .digest()
+      .slice(0, 16)
+    if (! bufferEqualConstantTime(flowSignature, expectedSignature)) {
+      return logInvalidContext(request, 'invalid signature')
+    }
+
+    log.info({
+      op: 'metrics.context.validate',
+      valid: true,
+      agent: request.headers['user-agent']
+    })
+    log.increment('metrics.context.valid')
+    return true
+  }
+
+  function logInvalidContext(request, reason) {
+    log.warn({
```

Comment:
Just out of interest, how do we decide whether something is `warn` or `error`?

Here for instance, the operation in question is failing but we only `warn`. Is that because said operation is not the main operation for an endpoint? Or are there other things that happen when the logs are processed that govern which log level is right?